src/celeritas/ext/detail/CelerOpticalPhysics.cc

```diff
@@ -0,0 +1,277 @@
+    params->SetVerboseLevel(options_.verbose);
+#endif
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build list of available particles.
+ */
+void CelerOpticalPhysics::ConstructParticle()
+{
+    // Eventually nothing to do here as Celeritas OpPhys won't generate
+    // G4OpticalPhotons
+    G4OpticalPhoton::OpticalPhotonDefinition();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build list of available processes and models.
+ */
+void CelerOpticalPhysics::ConstructProcess()
+{
+    auto* process_manager
+        = G4OpticalPhoton::OpticalPhoton()->GetProcessManager();
+    CELER_ASSERT(process_manager);
+
+    // Add Optical Processes
+    // TODO: Celeritas will eventually implement these directly (no
+    // G4OpticalPhotons) so how to set up on "Celeritas-side"
+    auto absorption = std::make_unique<G4OpAbsorption>();
+    if (process_is_active("OpAbsorption"))
+    {
+        process_manager->AddDiscreteProcess(absorption.release());
+        CELER_LOG(debug) << "Loaded Optical absorption with G4OpAbsorption "
+                            "process";
+    }
+
+    auto rayleigh = std::make_unique<G4OpRayleigh>();
+    if (process_is_active("OpRayleigh"))
+    {
+        process_manager->AddDiscreteProcess(rayleigh.release());
+        CELER_LOG(debug)
+            << "Loaded Optical Rayleigh scattering with G4OpRayleigh "
+               "process";
+    }
+
+    auto mie = std::make_unique<G4OpMieHG>();
+    if (process_is_active("OpMieHG"))
+    {
+        process_manager->AddDiscreteProcess(mie.release());
+        CELER_LOG(debug) << "Loaded Optical Mie (Henyey-Greenstein phase "
+                            "function) scattering with G4OpMieHG "
+                            "process";
+    }
+
+    // NB: boundary is also used later on in loop over particles,
+    // though it's only ever applicable to G4OpticalPhotons
+    auto boundary = std::make_unique<G4OpBoundaryProcess>();
+#if G4VERSION_NUMBER < 1070
+    boundary->SetInvokeSD(options_.invoke_sd);
+#endif
+    if (process_is_active("OpBoundary"))
+    {
+        process_manager->AddDiscreteProcess(boundary.get());
+        CELER_LOG(debug)
+            << "Loaded Optical boundary process with G4OpBoundaryProcess "
+               "process";
+    }
+
+    auto wls = std::make_unique<G4OpWLS>();
+#if G4VERSION_NUMBER < 1070
+    wls->UseTimeProfile(to_cstring(options_.wavelength_shifting));
+#endif
+    if (process_is_active("OpWLS"))
+    {
+        process_manager->AddDiscreteProcess(wls.release());
+        CELER_LOG(debug) << "Loaded Optical wavelength shifting with G4OpWLS "
+                            "process";
+    }
+
+#if G4VERSION_NUMBER >= 1070
+    auto wls2 = std::make_unique<G4OpWLS2>();
+    if (process_is_active("OpWLS2"))
+    {
+        process_manager->AddDiscreteProcess(wls2.release());
+        // I need to check how this differs from G4OpWLS...
```

Comment:
It looks identical except that it uses separate optical properties... so I guess it lets you have two different wavelength shifting processes in the same material?

Reply:
I did not realize WLS2 was akin to the scintillation components. That makes sense and we might consider generalizing it like we did with the scintillation components. 

PS: As an e.g., the NOvA detectors use a mineral oil with a dopant that scintillates and at least one other that produces a wavelength shifting. The scintillation/shifted photons get captured by an optical fiber that is also a wavelength-shifting fiber, so technically at least 2 WLS processes happen between scintillation and capture by the fiber that leads to a sensor.